src/Components/Blazor/Build/src/targets/Blazor.MonoRuntime.targets

```diff
@@ -219,49 +219,110 @@
     <WriteLinesToFile File="$(_BlazorLinkerOutputCache)" Lines="@(_LinkerResult)" Overwrite="true" />
   </Target>
 
+  <UsingTask TaskName="BlazorGetAssemblyReferences" AssemblyFile="$(BlazorTasksPath)" />
 
-  <UsingTask TaskName="ResolveBlazorRuntimeDependencies" AssemblyFile="$(BlazorTasksPath)" />
-  <Target
-    Name="_ResolveBlazorOutputsWhenNotLinked"
-    DependsOnTargets="_ResolveBlazorRuntimeDependencies"
-    Condition="'$(BlazorLinkOnBuild)' != 'true'">
+  <Target Name="_GetBlazorNotLinkedInputs">
+    <!--
+    We're using ILLink to determine the minimum closure of assemblies that need to be copied to the output directory. However executing ILLink results in a separate process being launched.
+    The outputs of a previous run are
+    -->
 
-    <ReadLinesFromFile File="$(_BlazorApplicationAssembliesCacheFile)" Condition="'@(_BlazorResolvedRuntimeDependencies->Count())' == '0'">
-      <Output TaskParameter="Lines" ItemName="_BlazorResolvedRuntimeDependencies"/>
-    </ReadLinesFromFile>
+    <PropertyGroup>
+      <_BlazorReferenceCacheFile>$(BlazorIntermediateOutputPath)blazor.assemblyreferences.cache</_BlazorReferenceCacheFile>
+    </PropertyGroup>
 
     <ItemGroup>
-      <BlazorOutputWithTargetPath Include="@(_BlazorResolvedRuntimeDependencies)">
-        <TargetOutputPath>$(BlazorRuntimeBinOutputPath)%(FileName)%(Extension)</TargetOutputPath>
-      </BlazorOutputWithTargetPath>
+      <_BlazorProjectReference Include="@(ReferencePath->WithMetadataValue('ReferenceSourceTarget', 'ProjectReference'))" />
+      <_BlazorExternalReference Include="@(ReferencePath)" Exclude="@(_BlazorProjectReference)" />
     </ItemGroup>
+
+    <BlazorGetAssemblyReferences Assemblies="@(IntermediateAssembly);@(_BlazorProjectReference)">
+      <Output TaskParameter="AssemblyReferences" ItemName="_ProjectAssemblyReferences" />
+    </BlazorGetAssemblyReferences>
+
+    <Hash ItemsToHash="@(_BlazorProjectReference);@(_ProjectAssemblyReferences);">
+      <Output TaskParameter="HashResult" PropertyName="_BlazorReferenceHash" />
+    </Hash>
+
+    <WriteLinesToFile
+      Lines="$(_BlazorReferenceHash)"
+      File="$(_BlazorReferenceCacheFile)"
+      Overwrite="True"
+      WriteOnlyWhenDifferent="True" />
   </Target>
 
   <Target
-    Name="_ResolveBlazorRuntimeDependencies"
+    Name="_GetBlazorNotLinkedAssemblyClosure"
+    DependsOnTargets="_GenerateBlazorLinkerDescriptor"
     Inputs="$(ProjectAssetsFile);
```

Comment:
This triggers incrementalism based on whether the .assets.json file modification timestamp has changed, right?

I assume you've checked that the file doesn't get touched in the case where its contents would be unchanged (for example, running `dotnet build` which internally does a restore).

Reply:
> I assume you've checked that the file doesn't get touched in the case where its contents would be unchanged (for example, running dotnet build which internally does a restore).

Yup, the file remains unchanged if it's contents remain the same. We use the same mechanism for MVC's assembly discovery task: https://github.com/aspnet/AspNetCore-Tooling/blob/master/src/Razor/src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.MvcApplicationPartsDiscovery.targets#L33